src/dkg/round2.rs

```diff
@@ -256,12 +264,86 @@ impl PublicPackage {
     }
 }
 
+/// A collection of [`PublicPackage`] structs, all from the same sender.
+#[derive(Default, Clone, PartialEq, Eq, Debug)]
+pub struct CombinedPublicPackage {
+    packages: Vec<PublicPackage>,
+}
+
+impl CombinedPublicPackage {
+    // This struct should not be constructed directly, hence `new` does not need to be `pub`.
+    // Keeping `new` private has the advantage that the implementation does not need to strictly
+    // enforce the same `sender_identity`, but it can omit this check (here we still check in debug
+    // builds just to catch bugs).
+    fn new(packages: Vec<PublicPackage>) -> Self {
+        // The serialization expects at least 1 package to be present
+        debug_assert!(!packages.is_empty());
+
+        let first_identity = &packages[0].sender_identity;
+        for pkg in &packages {
+            debug_assert_eq!(&pkg.sender_identity, first_identity);
+        }
+
+        Self { packages }
+    }
+
+    #[inline]
+    #[must_use]
+    pub fn packages(&self) -> &[PublicPackage] {
+        &self.packages
+    }
+
+    #[inline]
+    #[must_use]
+    pub fn packages_for<'a>(
+        &'a self,
+        recipient_identity: &'a Identity,
+    ) -> impl Iterator<Item = &'a PublicPackage> + 'a {
+        self.packages
+            .iter()
+            .filter(move |pkg| &pkg.recipient_identity == recipient_identity)
+    }
```

Comment:
should a `CombinedPublicPackage` ever have more than one package for a single recipient?

Reply:
I was initially thinking of this struct as something that anyone could construct. In that scenario, yes: it was possible. Now that I added restrictions (to `new` in particular) it should no longer be possible... unless somebody tampers with the serialization, but we can decide not to support that case.

I must say though that the round3 code wouldn't change much if we change this method to return an Option instead of an iterator. In fact, round3 would probably be a bit more verbose, so I'm not sure if it's worth changing.